File: DisasterServer/Entities/RMZShard.cs
using DisasterServer.Maps;
using DisasterServer.Session;
using DisasterServer.State;

namespace DisasterServer.Entities;

public class RMZShard : Entity
{
	private static byte GID;

	public byte ID = GID++;

	private bool _isSpawned;

	public RMZShard(int x, int y, bool spawned = false)
	{
		X = x;
		Y = y;
		_isSpawned = spawned;
	}

	public override TcpPacket? Destroy(Server server, Game game, Map map)
	{
		return null;
	}

	public override TcpPacket? Spawn(Server server, Game game, Map map)
	{
		return new TcpPacket(PacketType.SERVER_RMZSHARD_STATE, (byte)(_isSpawned ? 1u : 0u), ID, (ushort)X, (ushort)Y);
	}

	public override UdpPacket? Tick(Server server, Game game, Map map)
	{
		return null;
	}
}

File: disasterserver/entities/RMZShard.c
#include <entities/RMZShard.h>

bool shard_init(Server* server, Entity* entity)
{
	Shard* shard = (Shard*)entity;

	Packet pack;
	PacketCreate(&pack, SERVER_RMZSHARD_STATE);
	PacketWrite(&pack, packet_write8, shard->spawned);
	PacketWrite(&pack, packet_write16, shard->id);
	PacketWrite(&pack, packet_write16, (uint16_t)shard->pos.x);
	PacketWrite(&pack, packet_write16, (uint16_t)shard->pos.y);
	server_broadcast(server, &pack, true);

	return true;
}